xls/modules/rle/rle.x

```diff
@@ -0,0 +1,465 @@
+        input_r: chan<EncInData<SYMBOL_WIDTH>> in,
+        output_s: chan<EncOutData<SYMBOL_WIDTH, COUNT_WIDTH>> out,
+    ) {(input_r, output_s)}
+
+    next (tok: token, state: RLEEncState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+        let zero_input = EncInData { symbol: bits[SYMBOL_WIDTH]:0, last: false };
+        let (input_tok, input) = recv_if(tok, input_r, !state.prev_last, zero_input);
+
+        let prev_symbol_valid = state.prev_count != bits[COUNT_WIDTH]:0;
+        let symbol_differ = prev_symbol_valid && (input.symbol != state.prev_symbol);
+        let overflow = state.prev_count == std::unsigned_max_value<COUNT_WIDTH>();
+
+        let (symbol, count, last) = if (state.prev_last) {
+            (
+                bits[SYMBOL_WIDTH]:0,
+                bits[COUNT_WIDTH]:0,
+                false
+            )
+        } else if (symbol_differ || overflow) {
+            (
+                input.symbol,
+                bits[COUNT_WIDTH]:1,
+                input.last,
+            )
+        } else {
+            (
+                input.symbol,
+                state.prev_count + bits[COUNT_WIDTH]:1,
+                input.last,
+            )
+        };
+
+        let data = EncOutData {
+            symbol: state.prev_symbol,
+            count: state.prev_count,
+            last: state.prev_last
+        };
+
+        let do_send = state.prev_last || symbol_differ || overflow;
+        let data_tok = send_if(input_tok, output_s, do_send, data);
+
+        RLEEncState {
+            prev_symbol: symbol,
+            prev_count: count,
+            prev_last: last,
+        }
+    }
+}
+
+// RLE encoder specialization for the codegen
+proc RLEEnc32 {
+
+    init {()}
+
+    config (
+        input_r: chan<EncInData<32>> in,
+        output_s: chan<EncOutData<32, 2>> out,
+    ) {
+        spawn RLEEnc<u32:32, u32:2>(input_r, output_s);
+        ()
+    }
+
+    next (tok: token, state: ()) {
+        ()
+    }
+}
+
+// Tests
+
+const TEST_SYMBOL_WIDTH = u32:32;
+const TEST_COUNT_WIDTH = u32:2;
+
+//const MAX_COUNT = {std::unsigned_max_value<TEST_COUNT_WIDTH>()};
+const MAX_COUNT = u2:3;
+
+type TestSymbol = bits[TEST_SYMBOL_WIDTH];
+type TestCount = bits[TEST_COUNT_WIDTH];
+type TestEncInData = EncInData<TEST_SYMBOL_WIDTH>;
+type TestEncOutData = EncOutData<TEST_SYMBOL_WIDTH, TEST_COUNT_WIDTH>;
+
+// Tightly coupled testing approach
+
+#[test_proc]
+proc RLEEncTightTester {
+    terminator: chan<bool> out;
+    enc_input_s: chan<TestEncInData> out;
+    enc_output_r: chan<TestEncOutData> in;
+
+    init {()}
+    config (terminator: chan<bool> out) {
+        let (enc_input_s, enc_input_r) = chan<TestEncInData>;
+        let (enc_output_s, enc_output_r) = chan<TestEncOutData>;
+
+        spawn RLEEnc<TEST_SYMBOL_WIDTH, TEST_COUNT_WIDTH>(enc_input_r, enc_output_s);
+        (terminator, enc_input_s, enc_output_r)
+    }
+    next (tok: token, state:()) {
+
+        // Clear on `last`
+        let trans_send: TestEncInData[4] = [
```

Comment:
can we simplify the structure into a array? and craft the `TestEncInData` on the fly? (we set the the `last` flag when we're on the last element)